io-reactor/src/reactor.rs

```diff
@@ -248,62 +302,94 @@ impl<H: Handler, P: Poll> Runtime<H, P> {
                 .into();
 
             for res in self.listeners.values() {
-                self.poller.set_iterest(res, res.interests());
+                self.poller.set_interest(res, res.interests());
             }
             for res in self.transports.values() {
-                self.poller.set_iterest(res, res.interests());
+                self.poller.set_interest(res, res.interests());
             }
 
             // Blocking
-            let count = match self.poller.poll(Some(timeout)) {
+            #[cfg(feature = "log")]
+            log::trace!(target: "reactor", "Polling with timeout {timeout:?}");
+            match self.poller.poll(Some(timeout)) {
+                Ok(0) => {
+                    #[cfg(feature = "log")]
+                    log::trace!(target: "reactor", "Timeout");
+                    continue;
+                }
                 Ok(count) => count,
                 Err(err) => {
+                    #[cfg(feature = "log")]
+                    log::error!(target: "reactor", "Error during polling: {err}");
                     self.service.handle_error(err.into());
-                    0
+                    continue;
                 }
             };
 
             let instant = Instant::now();
             self.service.tick(instant);
 
-            if count > 0 {
-                self.handle_events(instant);
-            }
-            loop {
-                match self.cmd_recv.try_recv() {
-                    Err(chan::TryRecvError::Empty) => break,
-                    Err(chan::TryRecvError::Disconnected) => panic!("control channel is broken"),
-                    Ok(cmd) => self.service.handle_command(cmd),
+            let awoken = self.handle_events(instant);
+
+            self.handle_actions(instant);
```

Comment:
Can these lines be swapped to avoid binding to `awoken`?

Reply:
Alternatively maybe the below loops can be moved to `handle_events`?

Reply:
The lines can't be swapped since if we handle actions before the events the action queue will be empty and the actions generated by a service will not be handled until next loop

Reply:
Not sure I understand why you need commands to be handled before actions: this will result that in the case of shutdown we will not do pending actions from the service (for instance sending data to remote)